src/CallbackStream.php

```diff
@@ -0,0 +1,182 @@
+<?php
+/**
+ * Zend Framework (http://framework.zend.com/)
+ *
+ * @see       http://github.com/zendframework/zend-diactoros for the canonical source repository
+ * @copyright Copyright (c) 2015 Oscar Otero (http://oscarotero.com) / Zend Technologies USA Inc. (http://www.zend.com)
+ * @license   https://github.com/zendframework/zend-diactoros/blob/master/LICENSE.md New BSD License
+ */
+
+namespace Zend\Diactoros;
+
+use InvalidArgumentException;
+use RuntimeException;
+use Psr\Http\Message\StreamInterface;
+
+/**
+ * Implementation of PSR HTTP streams
+ */
+class CallbackStream implements StreamInterface
+{
+    /**
+     * @var callable|null
+     */
+    protected $callback;
+
+    /**
+     * @param callable $callback
+     * @throws InvalidArgumentException
+     */
+    public function __construct(callable $callback)
+    {
+        $this->attach($callback);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function __toString()
+    {
+        return $this->getContents();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function close()
+    {
+        $this->callback = null;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function detach()
+    {
+        $callback = $this->callback;
+        $this->callback = null;
+        return $callback;
+    }
+
+    /**
+     * Attach a new callback to the instance.
+     *
+     * @param callable $callback
+     * @throws InvalidArgumentException for callable callback
+     */
+    public function attach(callable $callback)
+    {
+        $this->callback = $callback;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getSize()
+    {
+    }
```

Comment:
The psr interface says:
```
/**
 * Get the size of the stream if known.
 *
 * @return int|null Returns the size in bytes if known, or null if unknown.
 */
```
So, it should be perfectly valid to return null. IMHO, this stream should not behave exactly like the standard stream (like you're proposing), because its goal is different. If I'll want to do what you say, there's no need of this class, because it can be done more easily:
```php
$result = $callable();
$response->getBody()->write($result);
```
The purpose of the callbackStream is to provide a way to send the response content in real time, so, you don't know the size or the time it will take. For example:
```php
$body = new CallbackStream(function () {
    echo 'The process is starting. Please, wait...';
    sleep(2);
    echo 'Running the process. Be patient, please';
    sleep(8);
    echo 'The process finished successfully';
});
```